torch/csrc/tensor/python_tensor.cpp

```diff
@@ -0,0 +1,261 @@
+
+void initialize_python_bindings(PyObject* module) {
+  // Initialize the at::Type* pointers, name, and properties of the PyTensorType
+  // vector. After this call, the vector must not be resized.
+  initialize_aten_types(tensor_types);
+
+  // Initialize the Python metaclass for the torch.Tensor, torch.FloatTensor,
+  // etc. types. The metaclass handles __instancecheck__ checks and binds the
+  // propeties is_cuda and is_sparse on the type objects.
+  py_initialize_metaclass(metaclass);
+
+  // Get the tp_dict of the Variable Python class. We copy function definitions
+  // onto each Tensor type object so that they can be accessed via e.g.
+  // `torch.Tensor.add`.
+  PyObject* var_dict = get_variable_dict();
+
+  // Initialize each Python type object torch.FloatTensor, torch.DoubleTensor,
+  // etc. and the "default" type object torch.Tensor.
+  for (auto& tensor_type : tensor_types) {
+    py_initialize_tensor_type(tensor_type.py_type, tensor_type.name, var_dict);
+  }
+
+  // The type object for torch.Tensor is at the end.
+  default_tensor_type = &tensor_types.back();
+
+  // Add the type objects to their corresponding modules. e.g. torch.FloatTensor
+  // is added to the `torch` module as `FloatTensor`. Also add all the type
+  // objects to the set torch._tensor_classes.
+  py_bind_tensor_types(tensor_types);
+}
+
+static void py_bind_tensor_types(const std::vector<PyTensorType>& tensor_types) {
+  auto torch_module = THPObjectPtr(PyImport_ImportModule("torch"));
+  if (!torch_module) throw python_error();
+
+  auto tensor_classes = THPObjectPtr(PyObject_GetAttrString(torch_module.get(), "_tensor_classes"));
+  if (!tensor_classes) throw python_error();
+
+  for (auto& tensor_type : tensor_types) {
+    auto name = std::string(tensor_type.name);
+    auto idx = name.rfind(".");
+    auto type_name = name.substr(idx + 1);
+    auto module_name = name.substr(0, idx);
+
+    auto module_obj = THPObjectPtr(PyImport_ImportModule(module_name.c_str()));
+    if (!module_obj) throw python_error();
+
+    PyObject* type_obj = (PyObject*)&tensor_type;
+    Py_INCREF(type_obj);
+    PyModule_AddObject(module_obj.get(), type_name.c_str(), type_obj);
+
+    if (PySet_Add(tensor_classes.get(), type_obj) < 0) {
+      throw python_error();
+    }
+  }
+}
+
+static bool PyTensorType_Check(PyObject* obj) {
+  auto it = std::find_if(tensor_types.begin(), tensor_types.end(),
+    [obj](const PyTensorType& x) {
+      return (PyObject*)&x == obj;
+    });
+  return it != tensor_types.end();
+}
+
+static at::Type* THPDefaultATenType;
+
+void set_default_tensor_type(const at::Type& type) {
+  set_type(*default_tensor_type, type.backend(), type.scalarType());
+  THPDefaultATenType = default_tensor_type->aten_type;
+}
+
+void py_set_default_tensor_type(PyObject* obj) {
+  if (!PyTensorType_Check(obj)) {
+    throw TypeError("invalid type object");
+  }
+  auto type = (PyTensorType*)obj;
+  if (!type->aten_type) {
+    throw TypeError("invalid type object");
```

Comment:
I think you need an error message as above, with special casing for good cuda error messages.  This can really happen because we create all of the "declared types".